source/m3_api_libc.c

```diff
@@ -105,6 +104,125 @@ void internal_itoa(int n, char s[], int radix)
     }
 }
 
+static
+void internal_uitoa(uint32_t n, char s[], int radix, bool upper)
+{
+    static char const HEXDIGITS_UPPER[0x10] = {
+            '0', '1', '2', '3', '4', '5', '6', '7',
+            '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'
+    };
+
+    static char const HEXDIGITS_LOWER[0x10] = {
+            '0', '1', '2', '3', '4', '5', '6', '7',
+            '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'
+    };
+
+    int i, j;
+    char c;
+
+    i = 0;
+    do {
+        s[i++] = (char)(upper ? HEXDIGITS_UPPER[n % radix] : HEXDIGITS_LOWER[n % radix]);
```

Comment:
```c
static char const HEXDIGITS[0x10] = {
  '0', '1', '2', '3', '4', '5', '6', '7',
  '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'
};
```

```suggestion
        char ch = HEXDIGITS[n % radix];  
        s[i++] = ch | (upper << 5);
```

Reply:
@MaxGraey That's for `%x` and `%X`. One is lower case and one is upper case. I may be able to do a plus op to use one `HEXDIGITS`, but I think this current implementation should also work fine?

Reply:
Oh nevermind, I misread your code... I will take that, thanks!

Reply:
it perform `X` -> `x` by `X | 32` -> `x`. With `upper`: `upper ? X : x | 32` or `ch | ((!upper) << 5)`

Reply:
> it perform X -> x by X | 32 -> x. With upper: upper ? X : X | 32 or X | ((!upper) << 5)

Yea sorry, I misread that earlier.

However, the memory corruption issue still exists. I'm still looking at it.